src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -103,8 +120,13 @@ func newMetricsAppender(pool *metricsAppenderPool) *metricsAppender {
 // reset is called when pulled from the pool.
 func (a *metricsAppender) reset(opts metricsAppenderOptions) {
 	a.metricsAppenderOptions = opts
-	if a.tagEncoder == nil {
-		a.tagEncoder = opts.tagEncoderPool.Get()
+
+	// Copy over any previous inuse encoders to the cached encoders list. If
+	// there are no cached encoders then create one.
+	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
+	a.inuseEncoders = a.inuseEncoders[:0]
```

Comment:
1) Need to nil out any refs before shrinking slice to `inuseEncoders[:0]` otherwise could hang onto it strongly forever.
2) Can we put this in a helper? I see this done other places later in the code.
```go
	a.cachedEncoders = append(a.cachedEncoders, a.inuseEncoders...)
	for i := range a.inuseEncoders {
		a.inuseEncoders[i] = nil
	}
	a.inuseEncoders = a.inuseEncoders[:0]
```

Reply:
done